// egg-errors/lib/framework/framework_base_error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameworkBaseError = exports.FRAMEWORK_ERROR_SYMBOL = void 0;
const __1 = require("../");
const formatter_1 = require("./formatter");
const assert = require("assert");
exports.FRAMEWORK_ERROR_SYMBOL = Symbol.for('FrameworkBaseError');
class FrameworkBaseError extends __1.EggBaseError {
    constructor(message, serialNumber, errorContext) {
        super({ message, serialNumber, errorContext });
        assert(message, 'message is required');
        assert(serialNumber, 'serialNumber is required');
        this.serialNumber = String(serialNumber);
        this.errorContext = errorContext || '';
        this.code = `${this.module}_${this.serialNumber}`;
        this[exports.FRAMEWORK_ERROR_SYMBOL] = true;
    }
    get module() {
        throw new Error('module should be implement');
    }
    // create a new frameworkError with format
    static create(message, serialNumber, errorContext) {
        const err = formatter_1.FrameworkErrorFormater.formatError(new this(message, serialNumber, errorContext));
        Error.captureStackTrace(err, this.create);
        return err;
    }
    static isFrameworkError(err) {
        return err[exports.FRAMEWORK_ERROR_SYMBOL] === true;
    }
}
exports.FrameworkBaseError = FrameworkBaseError;
